goethe/clean/cleaner.py

```diff
@@ -0,0 +1,78 @@
+import os
+import spacy
+
+LANG = 'de'
+BATCH_SIZE = 10000
+N_THREADS = 4
+
+
+class Cleaner:
+    def __init__(self, path):
+        self.path = path
+        self._nlp = None
+
+    def __iter__(self):
+        """Return iterator of cleaned sentences.
+        """
+        raise NotImplementedError
+
+    def write(self, write_path, delete=False):
+        """For a path 'pat/to/corpus' write:
+            path/to/corpus/
+                corpus.txt
+                corpus.tokens.txt
+        where 'corpus.txt' is one sent per line, spaces left in place,
+        and 'corpus.tokens.txt' is one sent per line, token separated by space.
+        """
+        folder, file, tokens = self._write_paths(write_path)
+        # Create corpus folder
+        os.makedirs(folder, exist_ok=True)
+
+        # Write sentence file
+        with open(file, 'w') as f:
+            for line in self:
+                f.write('%s\n' % line)
+
+        # Write tokens file
+        with open(tokens, 'w') as f:
+            for doc in self.tokenized_sents():
+                token_line = ' '.join(str(token) for token in doc)
+                f.write('%s\n' % token_line)
+
+        # Remove uncleaned data
+        if delete:
+            if os.path.isfile(self.path):
+                os.remove(self.path)
+            elif os.path.isdir(self.path):
+                os.rmdir(self.path)
+
+    def tokenized_sents(self):
+        """Return an iterator with SpaCy docs.
+        The docs' tokens can be iterated:
+        Example:
+            [token
+             for doc in self.tokenized_sents()
+             for token in doc]
+        """
+        return self.nlp.tokenizer.pipe(self,
```

Comment:
Does list generate a list of token lists, like this:
```python
sents = ["This is test one.", "This is test two!"]
...
tokens = [
 ["This", "is", "test", "one", "."],
 ["This", "is", "test", "two", "!"]
]
```
Or are the tokens generated by an generator while iterating? I ask this, because I think some tasks require the sentences to be a list of tokens.